### PR TITLE
Add `fisher_type="forward-only"` (FOOF method) option to `KFACLinearOperator`

### PR DESCRIPTION
**Pull request by runame**
Add a `fisher_type` to implement the preconditioner used for [FOOF](https://arxiv.org/abs/2201.12250). This represents an extreme value for this argument since we only do a forward pass, compute the input covariances, and set the gradient covariances to identity matrices.

### CODE REVIEW

File: curvlinops/kfac.py [hunk 1]
```diff
@@ -546,10 +560,23 @@ def _compute_loss_and_backward(self, output: Tensor, y: Tensor):
...
+            # we don't need to do a backward pass. See https://arxiv.org/abs/2201.12250.
+            # We choose to set the gradient covariance to the identity explicitly for
+            # the sake of simplicity, such that the rest of the code here and for
+            # `KFACInverseLinearOperator` does not have to be adapted. This could be
+            # changed to decrease the memory costs.
+            for mod_name, param_pos in self._mapping.items():
+                param = self._params[next(iter(param_pos.values()))]
+                self._gradient_covariances[mod_name] = eye(
+                    param.shape[0], dtype=param.dtype, device=self._device
+                )
```
Comment (f-dangel) [inline curvlinops/kfac.py:574]: Question: Can we simplify L571 or add a comment what it does? It seems to determine the output dimension, but I got slightly confused while reading it.

**Review by f-dangel — COMMENTED** [review]
Looks good, I have some minor comments.

One thing I wanted to propose is to add the following test:

Consider linear regression with square loss, `L =  R * \sum_n || W x_n - y_n ||^2` where `R` is the reduction factor from `MSELoss`. In this case, we have that `GGN(W) = 2 * FOOF(W) = 2 * I \otimes (R * \sum_n x_n x_n^T)`.

File: curvlinops/kfac.py [hunk 2]
```diff
@@ -546,10 +560,23 @@ def _compute_loss_and_backward(self, output: Tensor, y: Tensor):
             loss = self._loss_func(output, y)
             loss.backward()
 
+        elif self._fisher_type == "forward-only":
+            # Since FOOF sets the gradient covariance Kronecker factors to the identity,
+            # we don't need to do a backward pass. See https://arxiv.org/abs/2201.12250.
+            # We choose to set the gradient covariance to the identity explicitly for
+            # the sake of simplicity, such that the rest of the code here and for
+            # `KFACInverseLinearOperator` does not have to be adapted. This could be
+            # changed to decrease the memory costs.
```
[the remark this answers is not in the document]
Comment (f-dangel) [reply curvlinops/kfac.py:569]: I'm okay with that. Closing

**Comment by runame** [comment]
I like the proposed test and have just added it. Please check the math in the docstring.

File: test/kfac_cases.py [hunk 3]
```diff
@@ -128,3 +128,41 @@
...
+    },
+    # single-layer linear network with vector output
+    {
+        "model_func": lambda: Linear(5, 3),
+        "data": lambda: [
+            (rand(7, 5), regression_targets((7, 3))),
+            (rand(7, 5), regression_targets((7, 3))),
+        ],
+        "seed": 0,
+    },
```
Comment (f-dangel) [inline test/kfac_cases.py:156]: Could add a `Sequential(Conv2d, Flatten)` which also satisfies the test (I think!? Might have to double-check tomorrow as it is late already here).

File: test/test_kfac.py [hunk 4]
```diff
@@ -717,3 +717,157 @@ def test_logdet(case, exclude, separate_weight_and_bias, check_deterministic):
...
+    separate_weight_and_bias: bool,
+):
+    r"""Test KFAC with forward-only Fisher (FOOF) against exact GGN for one-layer model.
+
+    Consider linear regression with square loss, L =  R * \sum_n^N || W x_n - y_n ||^2,
+    where R is the reduction factor from the MSELoss. Per definition,
+    FOOF(W) = I \otimes (\sum_n x_n x_n^T / N). Hence, if R = 1 [reduction='sum'], we
+    have that GGN(W) = 2 * [I \otimes (\sum_n x_n x_n^T)] = 2 * N * FOOF(W).
+    If R = 1 / (N * C) [reduction='mean'], where C is the output dimension, we have
+    GGN(W) = 2 * R * [I \otimes (\sum_n x_n x_n^T)] = 2 / C * FOOF(W).
```
Comment (f-dangel) [inline test/test_kfac.py:820]: Looks correct to me.

**Review by f-dangel — APPROVED** [review]
Looks good to me!

File: test/kfac_cases.py (hunk 3 above)
Comment (runame) [reply test/kfac_cases.py:156]: I added a test for single-layer models with weight sharing (`Linear` and `Conv2d`). You can double check if the math makes sense to you (note the assumption of a mean reduction function for the reduce setting).

File: test/test_kfac.py [hunk 5]
```diff
@@ -717,3 +717,276 @@ def test_logdet(case, exclude, separate_weight_and_bias, check_deterministic):
...
+        # parameters are only initialized after the setting property is set
+        params = [p for p in model.parameters() if p.requires_grad]
+    data = data[setting]
+
+    # set appropriate loss_average argument based on loss reduction and setting
+    if loss_func.reduction == "mean":
+        if setting == "expand":
+            loss_average = "batch+sequence"
+        else:
+            loss_average = "batch"
```
Comment (f-dangel) [inline test/test_kfac.py:943]: Could compact into one line statement

File: test/test_kfac.py [hunk 6]
```diff
@@ -717,3 +717,276 @@ def test_logdet(case, exclude, separate_weight_and_bias, check_deterministic):
...
+    L =  R * \sum_n^N || W x_n - y_n ||^2, where R is the reduction factor from the
+    MSELoss. Per definition,
+    FOOF(W) = I \otimes (\sum_n (\sum_s^S x_{n,s} \sum_s^S x_{n,s}^T) / (N * S^2)),
+    where S is the weight-sharing dimension size. Hence, if R = 1 [reduction='sum'], we
+    have that
+    GGN(W) = 2 * [I \otimes (\sum_n \sum_s^S x_{n,s} \sum_s^S x_{n,s}^T) / S^2]
+    = 2 * N * FOOF(W) (assumes the mean/average pooling as reduction function).
+    If R = 1 / (N * C) [reduction='mean'], where C is the output dimension, we have
+    GGN(W) = 2 * R * [I \otimes (\sum_n \sum_s^S x_{n,s} \sum_s^S x_{n,s}^T) / S^2]
+    = 2 / C * FOOF(W) (assumes the mean/average pooling as reduction function).
```
Comment (f-dangel) [inline test/test_kfac.py:918]: Nit: Summations sometimes have start and/or end value, sometimes not.

File: test/test_kfac.py [hunk 7]
```diff
@@ -717,3 +717,276 @@ def test_logdet(case, exclude, separate_weight_and_bias, check_deterministic):
...
+    foof_mat = foof @ eye(foof.shape[1])
+
+    # Check for equivalence
+    num_data = sum(X.shape[0] for X, _ in data)
+    y: Tensor = data[0][1]
+    out_dim = y.shape[-1]
+    # See the docstring for the explanation of the scale
+    scale = num_data if loss_average is None else 1 / out_dim
+    if loss_average is None and setting == "expand":
+        X: Tensor = data[0][0]
```
Comment (f-dangel) [inline test/test_kfac.py:981]: Could combine with L976: `X, y = data[0]`.

**Review by f-dangel — APPROVED** [review]
Minor nits, feel free to merge! 👍
